## Author: Juan Francisco Alba Valle & Claudia Mu√±oz Mesa
## Contact: juanfav98@gmail.com/ claudiamzs22@gmail.com

## Script to determinate determinar the targets genes of transcriptional factor
## from  narrowPeak generated by MaCS2.


Args<-commandArgs(trailingOnly = T)
peak.col<- Args[[1]]
directory<- Args[[2]]

setwd(directory)

library("clusterProfiler")

library("org.At.tair.db")

library("topGO")

library("pathview")

library("ChIPseeker")

library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28


## Loadin-g peak files

col.peaks <- readPeakFile(peakfile = peak.col, header=FALSE)
col.peaks

## Defining the length of promoters

promoter <- getPromoters(TxDb=txdb, 
                         upstream=1000, 
                         downstream=1000)
promoter

## Annotation peaks

col.peak.annot<- annotatePeak(peak = col.peaks, 
                             tssRegion=c(-1000, 1000),
                             TxDb=txdb)
col.peak.annot

## Generating a plot with % of the peaks nature inside of DNA
## % exons, % promoters etc.

plotAnnoPie(col.peak.annot)
plotDistToTSS(col.peak.annot,
              title="Distribution of genomic loci relative to TSS",
              ylab = "Genomic Loci (%) (5' -> 3')")

## Generating a data frame with the annotation

control.annotation <- as.data.frame(col.peak.annot)
head(control.annotation)

target.genes.control <- unique(control.annotation$geneId[control.annotation$annotation == "Promoter"])
length(target.genes.control)

write(x = target.genes.control ,file = "target_genes_control.txt")

##################
## Analysis GEO ##
##################

# Defininf the universe to GEO anaylis

ath.genes <- as.data.frame(genes(txdb))
genes.chr1<- (ath.genes$gene_id)[ath.genes$seqnames == 1]

# ENRICHMENT GO

## Deininf KeyType

idType(OrgDb = org.At.tair.db)

## Kynd of analysis GEO
## "BP" biological process
## "MF" molecular function
## "CC" celular compartiment

e.bp<- enrichGO(gene = target.genes.control,
             OrgDb = org.At.tair.db,
             keyType = "TAIR",
             ont = "BP",
             pvalueCutoff = 0.05,
             qvalueCutoff = 0.05,
             universe = genes.chr1)

head(e.bp)

e.bp.table <- as.data.frame(e.bp)

head(e.bp.table)

barplot(e.bp, showCategory=20)


jpeg(filename = "plot_e_bp.jpeg", res = )
plot.e.bp<- plotGOgraph(x = e.bp,
                         firstSigNodes = 10,
                         useInfo = "all",
                         sigForAll = TRUE,
                         useFullNames = TRUE)
dev.off()

e.mf<- enrichGO(gene = target.genes.control,
                OrgDb = org.At.tair.db,
                keyType = "TAIR",
                ont = "MF",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                universe = genes.chr1)

head(e.mf)

e.mf.table <- as.data.frame(e.mf)

head(e.mf.table)

barplot(e.mf, showCategory=20)
 

jpeg(filename = "plot_e_mf.jpeg", width = 450 , height = 500)
plot.e.mf<- plotGOgraph(x = e.mf,
                         firstSigNodes = 10,
                         useInfo = "all",
                         sigForAll = TRUE,
                         useFullNames = TRUE)
dev.off()

e.cc<- enrichGO(gene = target.genes.control,
                OrgDb = org.At.tair.db,
                keyType = "TAIR",
                ont = "CC",
                pvalueCutoff = 0.05,
                qvalueCutoff = 0.05,
                universe = genes.chr1)

head(e.cc)

e.cc.table <- as.data.frame(e.cc)

head(e.cc.table)

barplot(e.cc, showCategory=20)
   

jpeg(filename = "plot_e_cc.jpeg")
plot.e.cc<- plotGOgraph(x = e.cc,
                         firstSigNodes = 10,
                         useInfo = "all",
                         sigForAll = TRUE,
                         useFullNames = TRUE)
dev.off()


# ENRICHMENT KEGG

e.kegg<- enrichKEGG(gene = target.genes.control, 
                    organism = "ath", 
                    universe = genes.chr1)

e.kegg.table <- as.data.frame(e.kegg)

head(e.kegg.table)

Kegg.ID<-e.kegg.table$ID


for(i in 1:length(Kegg.ID))
{
kegg.pathview<- pathview(gene.data = genes.chr1, 
                         pathway.id = Kegg.ID[i] ,
                         species = "ath",
                         gene.idtype = "TAIR")

}
